src/controllers/midi/midisourceclock.cpp

```diff
@@ -0,0 +1,117 @@
+#include "controllers/midi/midimessage.h"
+#include "controllers/midi/midisourceclock.h"
+#include "util/math.h"
+
+bool MidiSourceClock::handleMessage(unsigned char status) {
+    switch (status) {
+    case MIDI_START:
+        start();
+        return true;
+    case MIDI_STOP:
+        stop();
+        return true;
+    case MIDI_TIMING_CLK:
+        tick();
+        return true;
+    default:
+        return false;
+    }
+}
+
+void MidiSourceClock::start() {
+    m_bRunning = true;
+    m_iFilled = 0;
+    m_iRingBufferPos = 0;
+}
+
+void MidiSourceClock::stop() {
+    m_bRunning = false;
+}
+
+void MidiSourceClock::tick() {
+    // Update the ring buffer and calculate new bpm.  Update the last beat time
+    // if we are on a beat.
+
+    if (!m_bRunning) {
+        qDebug() << "MidiSourceClock: Got clock tick but not started, starting now.";
+        start();
+    }
+
+    // Ringbuffer filling.
+    // TODO(owen): We should have a ringbuffer convenience class.
+    const qint64 lastTickTime = m_pClock->now();
+    m_iTickRingBuffer[m_iRingBufferPos] = lastTickTime;
+    m_iRingBufferPos = (m_iRingBufferPos + 1) % kRingBufferSize;
+    if (m_iFilled < kRingBufferSize) {
+        ++m_iFilled;
+    }
+
+    // If this tick is a beat mark, record it, even if we have very few samples.
+    if (m_iRingBufferPos % kPulsesPerQuarter == 0) {
+        m_iLastBeatTime = lastTickTime;
+    }
+
+    // Figure out the bpm if we have enough samples.
+    if (m_iFilled > 2) {
+        qint64 earlyTickTime = 0;
+        if (m_iFilled < kRingBufferSize) {
+            earlyTickTime = m_iTickRingBuffer[0];
+        } else {
+            // In a filled ring buffer, the earliest tick is the next one that
+            // will get filled.
+            earlyTickTime = m_iTickRingBuffer[m_iRingBufferPos];
+        }
+        m_dBpm = calcBpm(earlyTickTime, lastTickTime, m_iFilled);
+    }
```

Comment:
Not really.  Midi is not precise enough to handle truly abrupt changes, and midi gear often gets out of sync when the tempo changes quickly, so that behavior is expected.  Just google "midi sync problems" sometime.  It's an old, imprecise standard.